cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -1365,16 +1398,51 @@ macro(build_brotli)
-                      BUILD_BYPRODUCTS "${BROTLI_STATIC_LIBRARY_ENC}"
-                                       "${BROTLI_STATIC_LIBRARY_DEC}"
-                                       "${BROTLI_STATIC_LIBRARY_COMMON}"
-                                       ${BROTLI_BUILD_BYPRODUCTS}
-                      CMAKE_ARGS ${BROTLI_CMAKE_ARGS}
-                      STEP_TARGETS headers_copy)
+  set(BROTLI_INSTALL_OVERRIDE)
+
+  if(CMAKE_SYSTEM_NAME STREQUAL "Emscripten")
+    # cmake install is disabled for brotli on emscripten, so we have
+    # to manually copy the libraries to our install directory
+    set(BROTLI_BUILD_DIR ${CMAKE_CURRENT_BINARY_DIR}/brotli_ep-prefix/src/brotli_ep-build)
+    set(BROTLI_BUILD_LIBS
+        "${BROTLI_BUILD_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}brotlienc-static${CMAKE_STATIC_LIBRARY_SUFFIX}"
+        "${BROTLI_BUILD_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}brotlidec-static${CMAKE_STATIC_LIBRARY_SUFFIX}"
+        "${BROTLI_BUILD_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}brotlicommon-static${CMAKE_STATIC_LIBRARY_SUFFIX}"
+    )
+
+    set(BROTLI_BUILD_INCLUDE_DIR
+        ${CMAKE_CURRENT_BINARY_DIR}/brotli_ep-prefix/src/brotli_ep/c/include/brotli)
+
+    externalproject_add(brotli_ep
+                        ${EP_COMMON_OPTIONS}
+                        URL ${BROTLI_SOURCE_URL}
+                        URL_HASH "SHA256=${ARROW_BROTLI_BUILD_SHA256_CHECKSUM}"
+                        BUILD_BYPRODUCTS "${BROTLI_STATIC_LIBRARY_ENC}"
+                                         "${BROTLI_STATIC_LIBRARY_DEC}"
+                                         "${BROTLI_STATIC_LIBRARY_COMMON}"
+                                         ${BROTLI_BUILD_BYPRODUCTS}
+                        CMAKE_ARGS ${BROTLI_CMAKE_ARGS}
+                        STEP_TARGETS headers_copy
+                        INSTALL_COMMAND "")
+    add_custom_command(TARGET brotli_ep
+                       POST_BUILD
+                       COMMAND ${CMAKE_COMMAND} -E copy_if_different
+                               ${BROTLI_BUILD_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}*${CMAKE_STATIC_LIBRARY_SUFFIX}
+                               ${BROTLI_LIB_DIR}
+                       COMMAND ${CMAKE_COMMAND} -E copy_directory
+                               ${BROTLI_BUILD_INCLUDE_DIR} ${BROTLI_INCLUDE_DIR}/brotli)
+  else() # not emscripten - just behave as normal
+    externalproject_add(brotli_ep
+                        ${EP_COMMON_OPTIONS}
+                        URL ${BROTLI_SOURCE_URL}
+                        URL_HASH "SHA256=${ARROW_BROTLI_BUILD_SHA256_CHECKSUM}"
+                        BUILD_BYPRODUCTS "${BROTLI_STATIC_LIBRARY_ENC}"
+                                         "${BROTLI_STATIC_LIBRARY_DEC}"
+                                         "${BROTLI_STATIC_LIBRARY_COMMON}"
+                                         ${BROTLI_BUILD_BYPRODUCTS}
+                        CMAKE_ARGS ${BROTLI_CMAKE_ARGS}
+                        STEP_TARGETS headers_copy)
+  endif()
```

Comment:
I'm trying this with `--preset=ninja-debug-emscripten`. I need at least the following on Debian GNU/Linux sid:

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index b50f3f1f1..9cd6e6727 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -936,11 +936,15 @@ set(EP_COMMON_TOOLCHAIN "-DCMAKE_C_COMPILER=${EP_C_COMPILER}"
                         "-DCMAKE_CXX_COMPILER=${EP_CXX_COMPILER}")
 
 if(CMAKE_AR)
-  list(APPEND EP_COMMON_TOOLCHAIN -DCMAKE_AR=${CMAKE_AR})
+  # Ensure using absolute path.
+  find_program(EP_CMAKE_AR ${CMAKE_AR} REQUIRED)
+  list(APPEND EP_COMMON_TOOLCHAIN -DCMAKE_AR=${EP_CMAKE_AR})
 endif()
 
 if(CMAKE_RANLIB)
-  list(APPEND EP_COMMON_TOOLCHAIN -DCMAKE_RANLIB=${CMAKE_RANLIB})
+  # Ensure using absolute path.
+  find_program(EP_CMAKE_RANLIB ${CMAKE_RANLIB} REQUIRED)
+  list(APPEND EP_COMMON_TOOLCHAIN -DCMAKE_RANLIB=${EM_CMAKE_RANLIB})
 endif()
 
 # External projects are still able to override the following declarations.
```

But I still get the following error:

```text
[34/34] Linking C executable brotli.js
FAILED: brotli.js 
: && /bin/emcc -Qunused-arguments -sUSE_ZLIB=1 -sSIDE_MODULE=1 -fPIC -fexceptions -fPIC -g -Werror -O0 -ggdb   -Wno-error  CMakeFiles/brotli.dir/c/tools/brotli.c.o -o brotli.js  libbrotlienc-static.a  libbrotlidec-static.a  libbrotlicommon-static.a && :
emcc: error: output suffix .js requested, but wasm side modules are just wasm files; emitting only a .wasm, no .js [-Wemcc] [-Werror]
ninja: build stopped: subcommand failed.
```

Do you know how to solve this?

Reply:
This one was debug only - SIDE_MODULE shouldn't be set on executable builds. Should be fixed now.